src/storage/v2/indices/text_index.cpp

```diff
@@ -0,0 +1,470 @@
+  }
+  return true;
+}
+
+bool TextIndex::DropIndex(std::string index_name) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  try {
+    mgcxx::text_search::drop_index(index_name);
+  } catch (const std::exception &e) {
+    throw query::QueryException(fmt::format("Tantivy error: {}", e.what()));
+  }
+  index_.erase(index_name);
+  std::erase_if(label_to_index_, [index_name](const auto &item) { return item.second == index_name; });
+  return true;
+}
+
+bool TextIndex::IndexExists(std::string index_name) const { return index_.contains(index_name); }
+
+std::vector<Gid> TextIndex::Search(std::string index_name, std::string search_query) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  if (!index_.contains(index_name)) {
+    throw query::QueryException(fmt::format("Text index \"{}\" doesn’t exist.", index_name));
+  }
+
+  auto input = mgcxx::text_search::SearchInput{.search_query = search_query, .return_fields = {"data", "metadata"}};
+  // Basic check for search fields in the query (Tantivy syntax delimits them with a `:` to the right)
+  if (search_query.find(":") == std::string::npos) {
+    input.search_fields = {"data"};
+  }
+
+  std::vector<Gid> found_nodes;
+  mgcxx::text_search::SearchOutput search_results;
+
+  try {
+    search_results = mgcxx::text_search::search(index_.at(index_name).context_, input);
+  } catch (const std::exception &e) {
+    throw query::QueryException(fmt::format("Tantivy error: {}", e.what()));
+  }
+  auto docs = search_results.docs;
+  for (const auto &doc : docs) {
+    // The CXX .data() method (https://cxx.rs/binding/string.html) may overestimate string length, causing JSON parsing
+    // errors downstream. We prevent this by resizing the converted string with the correctly-working .length() method.
+    auto doc_data = doc.data;
+    std::string doc_string = doc_data.data();
+    doc_string.resize(doc_data.length());
+    auto doc_json = nlohmann::json::parse(doc_string);
+    found_nodes.push_back(storage::Gid::FromString(doc_json["metadata"]["gid"].dump()));
+  }
+  return found_nodes;
+}
+
+void TextIndex::Commit() {
+  for (auto &[_, index_data] : index_) {
+    mgcxx::text_search::commit(index_data.context_);
+  }
+}
+
+void TextIndex::Rollback() {
+  for (auto &[_, index_data] : index_) {
+    mgcxx::text_search::rollback(index_data.context_);
+  }
+}
+
+std::vector<std::pair<std::string, LabelId>> TextIndex::ListIndices() const {
+  std::vector<std::pair<std::string, LabelId>> ret;
+  ret.reserve(index_.size());
+  for (const auto &[index_name, index_data] : index_) {
+    ret.push_back({index_name, index_data.scope_});
+  }
+  return ret;
+}
+
+std::uint64_t TextIndex::ApproximateVertexCount(std::string index_name) const { return 10; }
```

Comment:
Just checked – this method is a holdover from when I first made the class and it has no usages elsewhere. I’ll remove it since text indices aren’t to be used in planning/query optimization anyway.

Reply:
how they are not meant to be used with planning/query optimization? How do we know how many vertices will we fetch when inside query planner

Reply:
Label and label+property indices optimize looking up nodes with the indexed labels (and properties). Text indices are different in two things:
1. they don’t optimize node lookup
2. the only point when they’re searched is calls of the `text_search` query module
3. they optimize searching the contents of node properties

Because of (1) and (2), they don’t affect query plans and thus we don’t need the `ApproximateVertexCount` method here.